### PR TITLE
feat(remark-images): add option to wrap image with <figure> even when there is no caption

### LINKED ISSUE #19506: [gatsby-remark-images] Add new option to always wrap images in <figure>
## Summary

Allow to pass a new configuration option to the `gatsby-remark-images` plugin to wrap the HTML output in `<figure>` element, **independent of the existence of an image caption**.

### Basic example

At the moment, the `<figure>` element is rendered only whenever there is a caption:

https://github.com/gatsbyjs/gatsby/blob/38af21a17ed88554b1332c29dafd5dade4e9d7bd/packages/gatsby-remark-images/src/index.js#L333-L340

According to the [`<figure>` HTML documentation](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/figure), the `<figcaption>` is optional, so having only `<figure>` is semantically valid.

The option could be something like `wrapFigureOnlyWithCaptions` (_defaults to `true`_ for backwards compatibility).

### Motivation

The reason for this is mostly for styling purposes. In our case we want to style the `<figure>` element and have a consistent style not matter if there is a caption or not.

> I can open a PR and work on this if the new option is accepted.

### PR DESCRIPTION
**Pull request by emmenko**
## Description

Add option `wrapFigureOnlyWithCaptions`. By passing `false`, images without a caption will still be wrapped in a `<figure>` tag.

## Related Issues

Fixes #19506

### CODE REVIEW

File: packages/gatsby-remark-images/README.md [hunk 1]
```diff
@@ -44,21 +44,22 @@ plugins: [
...
+| `wrapFigureOnlyWithCaptions` | `true`  | Wrap the image with `<figure>` only when there is an image caption. By passing `false`, images without a caption will be wrapped in `<figure>` as well.                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |
```
Comment (emmenko) [inline packages/gatsby-remark-images/README.md]: This is the new option, the rest are formatting changes.

File: packages/gatsby-remark-images/src/index.js [hunk 2]
```diff
@@ -335,6 +335,12 @@ module.exports = (
   <figure class="gatsby-resp-image-figure" style="${wrapperStyle}">
     ${rawHTML}
     <figcaption class="gatsby-resp-image-figcaption">${imageCaption}</figcaption>
+  </figure>
```
Comment (pvdz) [inline packages/gatsby-remark-images/src/index.js]: Can we try to reduce duplication of markup?

How about something like this?

```suggestion
  if (imageCaption || !options.wrapFigureOnlyWithCaptions) {
    if (imageCaption) {
      rawHTML = `
        ${rawHTML}
        <figcaption class="gatsby-resp-image-figcaption">${imageCaption}</figcaption>
      `
    }
    rawHTML = `
      <figure class="gatsby-resp-image-figure" style="${wrapperStyle}">
        ${rawHTML}
      </figure>
    `
  }
```

Alternatively one back to back, less indentation.

File: yarn.lock [hunk 3]
```diff
@@ -21447,7 +21447,7 @@ type-of@^2.0.1:
   version "2.0.1"
   resolved "https://registry.yarnpkg.com/type-of/-/type-of-2.0.1.tgz#e72a1741896568e9f628378d816d6912f7f23972"
 
-typedarray-to-buffer@^3.1.5, typedarray-to-buffer@~3.1.5:
+typedarray-to-buffer@~3.1.5:
```
Comment (pvdz) [inline yarn.lock]: Please drop changes in this file. I think they don't need to be part of this commit.

File: packages/gatsby-remark-images/src/index.js (hunk 2 above)
Comment (emmenko) [reply packages/gatsby-remark-images/src/index.js]: I went with the if-else to try to make it more readable. I can change it as you suggested if you prefer so.

File: yarn.lock (hunk 3 above)
Comment (emmenko) [reply yarn.lock]: Sure, I just checked out the repo and run `yarn`, it wasn't intentional.

**Comment by emmenko** [comment]
@pvdz please have another look, I applied the changes as you suggested.

**Comment by pvdz** [comment]
Appologies. I shouldn't have stepped in. I missed that @blainekasten was already busy with this issue and should have let him review it.

**Comment by blainekasten** [comment]
@emmenko this PR is beautiful. You did a great job adding tests and keeping things non-breaking. I have good news and bad news. The good news is I have user-land solutions for you. The bad news is we are not going to accept this PR and i'll explain it all.

First why we do not want to accept this PR. We love community contributions and want to take them all. Unfortunately we are left with the burden of making sure our entire API surface is stable and doesn't break. One great way to do this is to be very intentional about APIs we expose. In react, props are a public API. While considering this PR it seems to be an issue that isn't likely to affect most users, *and* there are other solutions that you can implement yourself. So again, I'm really sorry that you did all of this great work and I hope you feel empowered to keep contributing, we love your work!

So the good news, there are a few solutions to this. I just spent about an hour diving in to make sure I understand how this whole system works. Notably there is a lot going on here and different setups can make certain things harder.

I'm not exactly sure what your setup is, but in general there are two ways this is going to be dealt with.

1. You are rendering markup via the htmlAst
2. You are rendering `gatsby-image` directly with the image served by remark.

So #2 has an interesting solution where basically if you have the underlying data, you can inspect it and see if there is a title. If there isn't, then you can manually wrap your `gatsby-image` call in a <figure>.

I think you are likely dealing with #1 based on some of the details you've shared. If this is the case, I think your best solution is some fun css. From what I can tell, you want to make sure your image + optional caption is always wrapped in a border.

I'm not sure exactly what css you'll need to do, but this is the sort of inspiration you could use:

```css
  .gatsby-resp-image-figure > .gatsby-resp-image-wrapper {
    border: none;
  }

  .gatsby-resp-image-figure,
  .gatsby-resp-image-wrapper {
    border: 2px solid red;
  }
```

Basically you're saying give a border to a figure or a wrapper, but if there is a figure and a child wrapper, don't border the wrapper, just the figure. I think that'll give you what you want (with some tweaks of course)

Please reach out if this doesn't solve your case, happy to help, but if we can avoid extra APIs that'd be our preference!

**Comment by emmenko** [comment]
Thanks again for your feedback, I understand the point of having a stable API.

Unfortunately as I mentioned in the issue the CSS hacking helps only until a certain point, that's why having a container element (`<figure>`) just solves the problem.

Of course I can write my own custom remark image plugin by copy/pasting the things that I need from the original plugin, but that's also not ideal in terms of maintenance.

I might have a new proposal though that is less opinionated and aims to provide a more generic extension point: would you consider adding a new option similar to what `wrapperStyle` does, which takes a function with the `rawHTML` as argument and returns a "final" `rawHTML` string?

The function can be used to perform some custom modifications to the markup, or in my case to wrap the `rawHTML` into a `<figure>` container.
The default implementation of the function is just an identity.
The new option could be named `wrapHtml` or whatever we come up with.

Let me know what you think, otherwise I'll try to find another solution. Thanks

**Comment by blainekasten** [comment]
Thanks for working with us here!

> Unfortunately as I mentioned in the issue the CSS hacking helps only until a certain point, that's why having a container element (<figure>) just solves the problem.

I'd like to understand what point it _doesn't work_. I understand it may not feel clean, but if it _can_ solve the problem, it seems like a good option until we hear that this is a common issue to warrant API changes.

> I might have a new proposal though that is less opinionated and aims to provide a more generic extension point: would you consider adding a new option similar to what wrapperStyle does, which takes a function with the rawHTML as argument and returns a "final" rawHTML string?

I think this is already possible without a new API. Mind showing me your component where you render the markdown html? Typically you have this `rawHTML` as a query in your component, where you could transform as a side effect in your component.

**Comment by emmenko** [comment]
> I'd like to understand what point it doesn't work.

I tried to explain it in the issue: https://github.com/gatsbyjs/gatsby/issues/19506#issuecomment-554050743
Basically when an image is "small" in width, having a full width frame is basically impossible without having a proper container element.

> I think this is already possible without a new API. Mind showing me your component where you render the markdown html? Typically you have this `rawHTML` as a query in your component, where you could transform as a side effect in your component.

Hmm we use the mdx plugin, so the images gets transformed there. I'm not sure it's possible to access the `rawHTML` from there. Or do you have a suggestion in that regards?
